### PR TITLE
Tanglescope/tanglewidthcollector: introduce new collector 

### PR DESCRIPTION
**Pull request by tsvisabo**

### CODE REVIEW

File: tanglescope/BUILD [hunk 1]
```diff
@@ -0,0 +1,73 @@
+cc_library(
+    name = "shared",
+    srcs = glob(
```
Comment (th0br0) [inline tanglescope/BUILD:3]: just do `*/**/*.cpp` and exclude tests?

File: tanglescope/BUILD [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    visibility = ["//visibility:public"],
+    deps = [
+        ":shared",
+    ],
+)
+
+cc_library(
+    name = "tanglewidthcollector",
+    srcs = ["tanglewidthcollector.cpp"],
+    hdrs = glob(["*.hpp"]),
```
Comment (th0br0) [inline tanglescope/BUILD]: fix

File: tanglescope/BUILD [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        "@com_github_gflags_gflags//:gflags",
+        "@com_github_google_glog//:glog",
+        "@prometheus_cpp",
+    ],
+)
+
+cc_library(
+    name = "tipselectioncollector",
+    srcs = ["tipselectioncollector.cpp"],
+    hdrs = glob(["tipselectioncollector.hpp"]),
```
Comment (th0br0) [inline tanglescope/BUILD]: no need for glob

File: tanglescope/BUILD [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        ":shared",
+        "//common/helpers",
+        "//common/trinary:tryte_long",
+    ],
+)
+
+cc_library(
+    name = "blowballcollector",
+    srcs = ["blowballcollector.cpp"],
+    hdrs = glob(["*.hpp"]),
```
Comment (th0br0) [inline tanglescope/BUILD]: fix

File: tanglescope/BUILD [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    visibility = ["//visibility:public"],
+    deps = [
+        ":shared",
+    ],
+)
+
+cc_library(
+    name = "echocollector",
+    srcs = ["echocollector.cpp"],
+    hdrs = glob(["*.hpp"]),
```
Comment (th0br0) [inline tanglescope/BUILD]: fix

File: tanglescope/collector/collector.cpp [hunk 6]
```diff
@@ -1,16 +1,13 @@
...
-        }
-        bool CollectorBase::parseConfiguration(const YAML::Node &conf) {
-            if (!conf){
-                return false;
-            }
-            return true;
-        }
-    }
+namespace tanglescope {
+CollectorBase::~CollectorBase() {}
```
Comment (th0br0) [inline tanglescope/collector/collector.cpp]: impl this in the header?

File: tanglescope/common/tangledb.cpp [hunk 7]
```diff
@@ -20,4 +21,29 @@ std::optional<TangleDB::TXRecord> TangleDB::find(const std::string& hash) {
...
+  auto now = std::chrono::system_clock::now();
+
+  std::unique_lock<std::shared_mutex> lock(mutex_);
+
+  std::vector<decltype(_txs)::key_type> vec;
+  for (auto&& it : _txs) {
+    if (std::chrono::duration_cast<std::chrono::seconds>(now -
+                                                         it.second.timestamp)
+            .count() > ageInSeconds) {
+      vec.emplace_back(it.first);
```
Comment (th0br0) [inline tanglescope/common/tangledb.cpp]: can't you just remove in-place?

File: tanglescope/common/zmqdbloader.cpp [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+void ZmqDBLoader::cleanDBPeriodically() {
+  auto pubThread = rxcpp::schedulers::make_new_thread();
+  auto pubWorker = pubThread.create_worker();
+
+  if (_cleanupInterval > 0) {
+    pubWorker.schedule_periodically(
+        pubThread.now(), std::chrono::seconds(_cleanupInterval),
+        [oldestTXAge = _oldestTXAge](auto scbl) {
+          TangleDB::instance().removeAgedTxs(oldestTXAge);
+          ;
```
Comment (th0br0) [inline tanglescope/common/zmqdbloader.cpp]: remove

File: tanglescope/common/zmqdbloader.hpp [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+ public:
+  void start();
+  bool parseConfiguration(const YAML::Node& conf);
+
+  constexpr static auto PUBLISHER = "publisher";
+  constexpr static auto OLDEST_TX_AGE = "oldest_tx_age";
+  constexpr static auto CLEANUP_INTERVAL = "cleanup_interval";
+
+  // typedef
+  typedef rxcpp::observable<std::shared_ptr<iri::IRIMessage>> ZmqObservable;
```
Comment (th0br0) [inline tanglescope/common/zmqdbloader.hpp]: you should use `using` syntax

File: tanglescope/tanglewidthcollector.cpp [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+void TangleWidthCollector::analyzeWidthPeriodically() {
+  auto pubThread = rxcpp::schedulers::make_current_thread();
+  auto pubWorker = pubThread.create_worker();
+
+  auto& thisRef = *this;
+  if (_snapshotInterval > 0) {
+    pubWorker.schedule_periodically(
+        pubThread.now() + std::chrono::seconds(_mesaureLineAge + 120),
+        std::chrono::seconds(_snapshotInterval),
+        [&thisRef](auto scbl) { thisRef.analyzeWidthImpl(); });
```
Comment (th0br0) [inline tanglescope/tanglewidthcollector.cpp]: This is a possible race condition: the class can be destructed before the thread is.
You should keep a reference to this thread & worker as a class member so you can destruct safely

Comment (th0br0) [reply tanglescope/tanglewidthcollector.cpp]: This might also apply to other collectors?

**Review by th0br0 — CHANGES_REQUESTED** [review]

File: tanglescope/common/tangledb.cpp (hunk 7 above)
Comment (tsvisabo) [reply tanglescope/common/tangledb.cpp]: that will invalidate the map it

File: tanglescope/common/zmqdbloader.hpp (hunk 9 above)
Comment (tsvisabo) [reply tanglescope/common/zmqdbloader.hpp]: i'll be using using

File: tanglescope/tanglewidthcollector.cpp (hunk 10 above)
Comment (tsvisabo) [reply tanglescope/tanglewidthcollector.cpp]: it might, though in practice tanglescope runner make sure not to exit the program before all collectors return, but we could implement a destructor to force exit

File: tanglescope/common/tangledb.cpp (hunk 7 above)
Comment (th0br0) [reply tanglescope/common/tangledb.cpp]: http://en.cppreference.com/w/cpp/algorithm/remove

File: .gitignore [hunk 11]
```diff
@@ -69,3 +69,5 @@ pom.xml
 
 /bazel*
 .bazelrc
+/cclient/bazel*
```
Comment (th0br0) [inline .gitignore]: just remove the `/` on line 70

File: tanglescope/common/zmqdbloader.hpp [hunk 12]
```diff
@@ -20,7 +20,7 @@ class ZmqDBLoader {
   constexpr static auto CLEANUP_INTERVAL = "cleanup_interval";
 
   // typedef
-  typedef rxcpp::observable<std::shared_ptr<iri::IRIMessage>> ZmqObservable;
+  using ZmqObservable = rxcpp::observable<std::shared_ptr<iri::IRIMessage>> ;
```
Comment (th0br0) [inline tanglescope/common/zmqdbloader.hpp]: remove the whitespace

File: tanglescope/runner/configuration.yaml [hunk 13]
```diff
@@ -3,7 +3,7 @@ dbloader:
   #URL of ZMQ publisher to connect to
   publisher: "tcp://zmq.testnet.iota.org:5556"
   #TX's older than that age will be removed each cleanup_interval [seconds]
-  oldest_tx_age: 600
+  oldest_tx_age: 1800
```
Comment (th0br0) [inline tanglescope/runner/configuration.yaml]: is this change on purpose?

File: tanglescope/tanglewidthcollector.cpp [hunk 14]
```diff
@@ -41,48 +41,71 @@ void TangleWidthCollector::collect() {
...
+                                                {4 * MEASURE_LINE_AGE_STEP, 0}};
 
-  for (auto&& kv : txs) {
-    if (!isOlderThanMeasureLine(now, kv.second.timestamp)) {
-      auto trunkIt = txs.find(kv.second.trunk);
-      if (trunkIt != txs.end()) {
-        if (isOlderThanMeasureLine(now, trunkIt->second.timestamp)) {
-          ++width;
+  auto now = std::chrono::system_clock::now();
+  auto measureLine =
```
Comment (th0br0) [inline tanglescope/tanglewidthcollector.cpp]: what's the motivation here?

Comment (th0br0) [reply tanglescope/tanglewidthcollector.cpp]: could you add some in-code documentation or comment the method?

**Review by th0br0 — CHANGES_REQUESTED** [review]

File: tanglescope/runner/configuration.yaml (hunk 13 above)
Comment (tsvisabo) [reply tanglescope/runner/configuration.yaml]: yes, i want the db to be consistent enough

**Review by th0br0 — APPROVED** [review]
